File: java/src/main/java/adventofcode/day3/TobogganTrajectory.java
package adventofcode.day3;

import javafx.util.Pair;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

// Day 3 of Advent of Code challenge
// Link: https://adventofcode.com/2020/day/3
public class TobogganTrajectory {

  private static final char TREE = '#';

  // Used in part 2
  private static final List<Pair<Integer, Integer>> slopes = new ArrayList<Pair<Integer, Integer>>() {{
    add(new Pair<>(1, 1));
    add(new Pair<>(3, 1));
    add(new Pair<>(5, 1));
    add(new Pair<>(7, 1));
    add(new Pair<>(1, 2));
  }};

  public static int solution(int part, Pair<Integer, Integer> slope) {
    List<List<Character>> grid = parseInput();
    int row = 0;
    int col = 0;
    int result = 0;

    while (row < grid.size()) {
      if (grid.get(row).get(col) == TREE) result++;
      int nextCol = part == 1 ? col + 3 : col + slope.getKey();
      row = part == 1 ? row + 1 : row + slope.getValue();
      col = nextCol % grid.get(0).size();
    }

    return result;
  }

  private static List<List<Character>> parseInput() {
    List<List<Character>> grid = new ArrayList<>();

    try {
      FileReader fileReader = new FileReader("java/src/main/resources/inputs/day3.txt");
      BufferedReader bufferedReader = new BufferedReader(fileReader);
      String line = bufferedReader.readLine();

      while (line != null) {
        List<Character> row = new ArrayList<>();
        for (char c : line.toCharArray()) row.add(c);
        grid.add(row);
        line = bufferedReader.readLine();
      }
    } catch (IOException e) {
      e.printStackTrace();
    }

    return grid;
  }

  public static void main(String[] args) {
    System.out.println("Part 1: Number of trees: " + solution(1, null));
    int part2Result = 1;
    for (Pair<Integer, Integer> slope : slopes) {
      part2Result *= solution(2, slope);
    }
    System.out.println("Part 2: Number of trees: " + part2Result);
  }
}

File: golang/day3/toboggan_trajectory.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

// Day 3 of Advent of Code challenge
// Link: https://adventofcode.com/2020/day/3

// Slope to represent slope values for part 2 solution
type Slope struct {
	x int
	y int
}

func parseInput() []string {
	var grid []string
	inputFile, error := os.Open("input.txt")
	if error != nil {
		log.Fatal(error)
	}
	defer inputFile.Close()

	scanner := bufio.NewScanner(inputFile)
	for scanner.Scan() {
		grid = append(grid, scanner.Text())
	}

	if error := scanner.Err(); error != nil {
		log.Fatal(error)
	}

	return grid
}

func solution1() int {
	var grid []string = parseInput()
	return countTrees(3, 1, grid)
}

func solution2() int {
	var slopes [5]Slope = [5]Slope{{1, 1}, {3, 1}, {5, 1}, {7, 1}, {1, 2}}
	var grid []string = parseInput()
	var result int = 1

	for _, slope := range slopes {
		result *= countTrees(slope.x, slope.y, grid)
	}

	return result
}

func countTrees(slopeX, slopeY int, grid []string) int {
	var row int = 0
	var col int = 0
	var numTrees int = 0

	for row < len(grid) {
		if grid[row][col] == '#' {
			numTrees++
		}
		col = (col + slopeX) % len(grid[row])
		row = row + slopeY
	}

	return numTrees
}

func main() {
	var result1 = solution1()
	var result2 = solution2()
	fmt.Printf("Part 1: Number of trees: %d", result1)
	fmt.Println("")
	fmt.Printf("Part 2: Number of trees: %d", result2)
}
